packages/injectable/ogre-tools-injectable.d.ts

```diff
@@ -1,150 +1,133 @@
-        di: DiContainerForInstantiate,
-        ...instantiationParameter: TentativeTuple<
-          InferFromInjectable<TInjectable>[1]
-        >
-      ) => InferFromInjectable<TInjectable>[0],
+    override<
+      Instance extends InjectionInstance,
+      InjectionInstance,
+      InstantiationParam
+    >(
+      injectable: Injectable<Instance, InjectionInstance, InstantiationParam>,
+      instantiateStub: Instantiate<Instance, InstantiationParam>,
     ): void;
 
-    register(injectable: NormalInjectable<any, any>): void;
-    register(injectable: InjectionTokenInjectable<any>): void;
+    register<
+      InjectionInstance extends Instance,
+      Instance,
+      InstantiationParam
+    >(injectable: Injectable<InjectionInstance, Instance, InstantiationParam>): void;
     preventSideEffects: () => void;
   }
 
-  export interface DiContainerForSetup extends DiContainerForInjection<true> {}
+  export type Instantiate<Instance, InstantiationParam> = {
+    (di: DiContainerForInstantiate, param: InstantiationParam extends void
+      ? void
+      : InstantiationParam
+    ): Instance;
+  };
 
-  export interface DiContainerForInstantiate
-    extends DiContainerForInjection<false> {}
+  export type DiContainerForSetup = DiContainerForInjection<true>;
+  export type DiContainerForInstantiate = DiContainerForInjection<false>;
 
-  type InjectionToken<TInstance, TInstantiationParameter> = {
-    template: TInstance;
-    instantiationParameter: TInstantiationParameter;
+  export interface InjectionToken<
+    Instance,
+    InstantiationParam,
+  > {
+    template: Instance;
+    instantiationParameter: InstantiationParam;
     key: Symbol;
-  };
+  }
 
-  interface CommonInjectable {
+  export interface Injectable<
+    Instance extends InjectionInstance,
+    InjectionInstance,
+    InstantiationParam,
+  > {
     id: string;
     setup?: (di: DiContainerForSetup) => void | Promise<void>;
     causesSideEffects?: boolean;
-    lifecycle?: ILifecycle;
-  }
-
-  interface InjectionTokenInjectable<
-    TInjectionToken extends InjectionToken<unknown, unknown>,
-  > extends CommonInjectable {
-    injectionToken: TInjectionToken;
-
-    instantiate: (
-      di: DiContainerForInstantiate,
-      instantiationParameter: InferFromToken<TInjectionToken>[1],
-    ) => InferFromToken<TInjectionToken>[0];
-  }
-
-  interface NormalInjectable<TInstance, TInstantiationParameter>
-    extends CommonInjectable {
-    instantiate: (
-      di: DiContainerForInstantiate,
-      instantiationParameter: TInstantiationParameter,
-    ) => TInstance;
+    injectionToken?: InjectionToken<InjectionInstance, InstantiationParam>;
+    instantiate: Instantiate<Instance, InstantiationParam>;
+    lifecycle: ILifecycle<
+      InstantiationParam,
+      string | number | symbol
+    >;
   }
 
   export function getInjectable<
-    TInjectionToken extends InjectionToken<unknown, unknown>,
+    Instance extends InjectionInstance,
+    InjectionInstance,
```

Comment:
This requires better naming of type parameters to communicate intent. `TSubType` and `TSuperType`? Or `TInjectableInstance` and `TInjectionTokenInstance`?

Reply:
I think `InjectionTokenInstance` is best, will change.